File: src/leetcode/CircularArrayLoop.java
package leetcode;
import java.util.*;

public class CircularArrayLoop {
    public static boolean circularArrayLoop(int[] nums) {
    	int len = nums.length;
    	for(int i=0; i<len; i++) {
    		int startPosition = i;
    		int direction = (nums[startPosition]>0)? 1 : -1; //-1 backward, 1 forward
    		LinkedList<Integer> visited = new LinkedList<>();
    		LinkedList<Integer> travel = new LinkedList<>();
    		
    		visited.offer(startPosition);
    		travel.offer(startPosition);
    		while(!travel.isEmpty()) {
    			int location = travel.poll();
    			int value = nums[location];
    			int flag = (value>0)? 1 : -1;
    			if(flag != direction)
    				break;
    			startPosition = (startPosition+value)%len;
    			if(startPosition < 0)
    				startPosition += len;
    			if(visited.contains(startPosition)){
    				if(visited.get(visited.size()-1) != startPosition)
    				    return true;
    				else
    					break;
    			}
    			visited.offer(startPosition);
    			travel.offer(startPosition);
    			
    		}
    	}
        return false;
    }
    public static void main(String[] args) {
    	int[] nums = new int[]{-1, -2, -3, -4, -5};
    	boolean result = circularArrayLoop(nums);
    	System.out.println(result);
    }
}

File: src/leetcode/CircularArrayLoop.py
class Solution(object):
    def circularArrayLoop(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        length = len(nums);
        for i in range(length):
            startPosition = i;
            direction = 1 if nums[i]>0 else -1
            visited = [i]
            travel = [i]
            while travel:
                location = travel.pop(0)
                value = nums[location]
                flag = 1 if value>0 else -1
                if not direction== flag:
                    break
                startPosition = (startPosition+value)%length
                if startPosition < 0:
                    statrPosition += length
                if startPosition in visited:
                    if not visited[-1] == startPosition:
                        return True
                    else:
                        break
                visited.append(startPosition)
                travel.append(startPosition)
        return False


if __name__ == "__main__":
    nums = [-1, 2]
    solution = Solution()
    result = solution.circularArrayLoop(nums)
    print result
